### DISCUSSION TITLE
Model Conversion Guide | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#1904] [Q&A]
# Model Conversion Guide | Seeed Studio Wiki

Model Conversion

https://wiki.seeedstudio.com/recamera_model_conversion/

**Comment by O-J1** [comment]
I eagerly await this as the product is not really useful without this.

**Comment by mike-ensor** [comment]
Is there a timeline on when/how we can upload or replace a model on the camera?

**Comment by afrapic** [comment]
Ok, first of all, installation with method 1 gives a lot of errors, seems that is failing but ends up running anyways. Weird. Second, conversion with the onnx model you provided works, but i'm trying with the yolov8n and it doesn't work. I changed the output names, seem to be different. It breaks in the middle trying to run an intermediate onnx file. How the onnx conversion is done? Evidently, not with the export command of Ultralytics. Some info on this? Thanks

**Reply by jjjadand** [reply]
hi, afrapic. Due to environmental issues, we recommend performing model conversion in a Docker container (method 2). The yolo11n.onnx in the link is indeed obtained using Ultralytics' API. We have tested the deployment of yolov8n, and it works fine. Could you share your conversion command and the error info?
We are collaborating with Ultralytics to make the deployment of YOLO simpler and more detailed in the future. Thank you!

**Reply by afrapic** [reply]
Thanks! I managed to find out what was the problem: i was doing the export on my windows machine, and trying the conversion in WSL Ubuntu with a different ONNX version. If I do the export in the WSL Ubuntu machine now I have some progress. Still can't finish the conversion. I'm trying to convert to INT8, with the docker now, as suggested. I managed to go up to the calibration table stage, but then when i go to the last step it stops here

```bash
[CMD]: model_runner.py --input yolov8n_in_ori.npz --model yolov8n_cv181x_int8_sym_tpu.mlir --output yolov8n_cv181x_int8_sym_tpu_outputs.npz
[                                                  ] 0%
```
Could it be that I need a full local Ubuntu machine and not the WSL Ubuntu?

**Reply by jjjadand** [reply]
Yes, our tests are based on local Ubuntu, and WSL Ubuntu may have potential issues. It seems to be stuck at the inference testing during the calibration process.

**Reply by afrapic** [reply]
Thank you so much for all the info! Really helpful. I'll see if i can find a local Ubuntu. I'm working in my company issued laptop, which only runs Windows.

**Reply by tuladhar2002** [reply]
I was running into same issue too while running local Ubuntu 22.04 as suggested. The model_runner kept stalling at 0%. 
I tried removing the  --test_input argument from the model_deploy command and it seemed to work for me. Still don't really know why it kept stalling like that, but just remove that argument for the int8 quantization and should work.

**Comment by moxilia** [comment]
I have already converted my trained YOLO11 model to CVImodel. First, I converted it to ONNX, then to MLIR. However, when I load the model in Node-RED, I get an error saying that the model is not supported. I have checked that the ONNX versions match the ones recommended in the wiki. Can anyone help me?

**Reply by jjjadand** [reply]
> I have already converted my trained YOLO11 model to CVImodel. First, I converted it to ONNX, then to MLIR. However, when I load the model in Node-RED, I get an error saying that the model is not supported. I have checked that the ONNX versions match the ones recommended in the wiki. Can anyone help me?

hi, @moxilia . You've got the `.cvimodel` file, not the intermediate `.milr` file from the conversion? Could you share your .cvimodel file with me to check?

**Reply by 3DSBricker** [reply]
@jjjadand Hi, I've got the same problem. I have tried the Int8 and F16 model. I have uploaded the correct cvimodel. Still, it says model not supported. Any idea? 
Since you asked, here are both of my attempts:

[CVImodels.zip](https://github.com/user-attachments/files/20477064/CVImodels.zip)

**Reply by jjjadand** [reply]
@3DSBricker hi, could u share the step of conversation and your onnx model? thanks

**Reply by takespace** [reply]
Hi @jjjadand , I have same issue.
I could .cvimodel but I got "Model not supported" after I uploaded the model on Node-Red.
I uploaded 2 file converted from yolo11n.onnx model provided in the instuction page.
Do you have any idea?
[yolo11n.zip](https://github.com/user-attachments/files/21455408/yolo11n.zip)

**Reply by litxaohu** [reply]
First, you need to make sure that you are using the yolo11 or yolo8 version of the model for the conversion. If this is not a problem, I think it might be a version issue with the tpu_mlir library. We have tested version 1.7 and found no issues. The original 1.14 version would cause problems where the model couldn't be recognized. You can refer to the latest wiki document:
https://wiki.seeedstudio.com/recamera_model_conversion/#convert-and-quantize-ai-models-to-the-cvimodel-format
<img width="1644" height="203" alt="image" src="https://github.com/user-attachments/assets/4438d01a-1225-47d1-94b2-07f437f770ec" />

**Reply by takespace** [reply]
@litxaohu 
Thank you for your advice. The issue has been successfully resolved. I was using this site as a reference: https://wiki.seeedstudio.com/ja/recamera_model_conversion/. It seems the information there was for v1.4, which is why it didn't work.

**Comment by matsujirushi** [comment]
I tried to convert yolo11n.onnx to INT8 as per the instructions, but the error occurred due to version differences between the tpu_mlir package and the sophgo/tpu-mlir repository.

The complete script reviewed is shown below.

* tpuc_dev container v3.3(latest)
* tpu_mlir v1.16

```
docker run --privileged --name MyName -v $PWD:/workspace -it sophgo/tpuc_dev:v3.3

git clone -b v1.16 --depth 1 https://github.com/sophgo/tpu-mlir.git
cd tpu-mlir
./release_pip.sh
pip install ./dist/tpu_mlir-1.16-py3-none-any.whl[all]

mkdir model_yolo11n && cd model_yolo11n
cp -rf ../regression/dataset/COCO2017 .
cp -rf ../regression/image .
mkdir Workspace && cd Workspace

# cp /workspace/yolo11n.onnx .

model_transform   --model_name yolo11n   --model_def yolo11n.onnx   --input_shapes "[[1,3,640,640]]"   --mean "0.0,0.0,0.0"   --scale "0.0039216,0.0039216,0.0039216"   --keep_aspect_ratio   --pixel_format rgb   --output_names "/model.23/cv2.0/cv2.0.2/Conv_output_0,/model.23/cv3.0/cv3.0.2/Conv_output_0,/model.23/cv2.1/cv2.1.2/Conv_output_0,/model.23/cv3.1/cv3.1.2/Conv_output_0,/model.23/cv2.2/cv2.2.2/Conv_output_0,/model.23/cv3.2/cv3.2.2/Conv_output_0"   --test_input ../image/dog.jpg   --test_result yolo11n_top_outputs.npz   --mlir yolo11n.mlir

run_calibration   yolo11n.mlir   --dataset ../COCO2017   --input_num 100   -o yolo11n_calib_table

model_deploy   --mlir yolo11n.mlir   --quantize INT8   --quant_input   --processor cv181x   --calibration_table yolo11n_calib_table     --test_reference yolo11n_top_outputs.npz   --customization_format RGB_PACKED   --fuse_preprocess   --aligned_input   --model yolo11n_1684x_int8_sym.cvimodel
```

The last command removes the `--test_input` option. See [here](https://github.com/Seeed-Studio/wiki-documents/discussions/1904#discussioncomment-12671229).

**Reply by jjjadand** [reply]
`tpuc_dev:v3.3` and `tpu_mlir v1.16` have not been fully tested and have some bugs.
Currently, it is recommended to use `tpuc_dev:v3.1` and `tpu_mlir v1.14`.

**Comment by matsujirushi** [comment]
I think the image is incorrect.

![image](https://github.com/user-attachments/assets/f3f4aedf-08be-4707-9c4f-fe8420ef40a9)

**Reply by jjjadand** [reply]
> I think the image is incorrect.
> 
> ![image](https://private-user-images.githubusercontent.com/8079404/429452079-f3f4aedf-08be-4707-9c4f-fe8420ef40a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM5OTAyNTEsIm5iZiI6MTc0Mzk4OTk1MSwicGF0aCI6Ii84MDc5NDA0LzQyOTQ1MjA3OS1mM2Y0YWVkZi0wOGJlLTQ3MDctOWM0Zi1mZTg0MjBlZjQwYTkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA0MDdUMDEzOTExWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmE3NWUxYzNkZjQ4MzdlNjM5ZjM2M2M3YWE1MjlhZWMxZjQ4ZjMwN2M0MmZkYmZiOWVmZjhmNjQyMzQ4NmVjMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.BpOywdBiEZdPVzqooHR29uQ0UKCeM9dNyhAJQ0hybXU)

hi, matsujirushi. The input names of the concat operator correspond to the output names before head. Could you help point out where the mistake is? Thanks.

**Reply by matsujirushi** [reply]
Hi @jjjadand ,
There are three images in the `--output_names` option description.

/model.23/cv2.0/cv2.0.2/Conv_output_0
/model.23/cv3.0/cv3.0.2/Conv_output_0

-> **No image.**

/model.23/cv2.1/cv2.1.2/Conv_output_0
/model.23/cv3.1/cv3.1.2/Conv_output_0

-> 2nd image.

/model.23/cv2.2/cv2.2.2/Conv_output_0
/model.23/cv3.2/cv3.2.2/Conv_output_0

-> 1st and 3rd image.

**Reply by jjjadand** [reply]
> Hi @jjjadand , There are three images in the `--output_names` option description.
> 
> /model.23/cv2.0/cv2.0.2/Conv_output_0 /model.23/cv3.0/cv3.0.2/Conv_output_0
> 
> -> **No image.**
> 
> /model.23/cv2.1/cv2.1.2/Conv_output_0 /model.23/cv3.1/cv3.1.2/Conv_output_0
> 
> -> 2nd image.
> 
> /model.23/cv2.2/cv2.2.2/Conv_output_0 /model.23/cv3.2/cv3.2.2/Conv_output_0
> 
> -> 1st and 3rd image.

Yes, if the outputs before the `head` are used as parameters, then `--output_names` will include six outputs. If the final output of the model is used, then `--output_names` will include only one output. Because of the sscma's API in reCamera, YOLO11 uses the former approach to choose the parameters for `--output_names`.

**Reply by matsujirushi** [reply]
I don't see any image as the basis for `/model.23/cv2.0/cv2.0.2/Conv_output_0 /model.23/cv3.0/cv3.0.2/Conv_output_0`, is this correct?

**Reply by jjjadand** [reply]
> I don't see any image as the basis for `/model.23/cv2.0/cv2.0.2/Conv_output_0 /model.23/cv3.0/cv3.0.2/Conv_output_0`, is this correct?

Thank u. You helped identify an error. The first image should correspond to `/model.23/cv2.0/cv2.0.2/Conv_output_0` and `/model.23/cv3.0/cv3.0.2/Conv_output_0`, but it was incorrectly linked to other images during editing. we will fix the figure.

**Comment by psawai** [comment]
I have got the FP16 cvimodel for YOLO11. However, when I tried to get the INT8 cvimodel, I was unable to succeed. The command line execution seemed to terminate without completing.
The execution process is as follows:
root@85cdfc2a6825:/workspace/mywork/workspace/model_yolo11n/Workspace# model_deploy --mlir yolo11n.mlir --quantize INT8 --quant_input --processor cv181x --calibration_table yolo11n_calib_table --test_input ../image/dog.jpg --test_reference yolo11n_top_outputs.npz --customization_format RGB_PACKED --fuse_preprocess --aligned_input --model yolo11n_1684x_int8_sym.cvimodel
2025/04/22 10:37:08 - INFO : TPU-MLIR v1.14-20241231
2025/04/22 10:37:08 - INFO :
  load_config Preprocess args :
        resize_dims           : [640, 640]
        keep_aspect_ratio     : True
        keep_ratio_mode       : letterbox
        pad_value             : 0
        pad_type              : center
        input_dims            : [640, 640]
        --------------------------
        mean                  : [0.0, 0.0, 0.0]
        scale                 : [0.0039216, 0.0039216, 0.0039216]
        --------------------------
        pixel_format          : rgb
        channel_format        : nchw

2025/04/22 10:37:08 - INFO : Add preprocess, set the following params:
2025/04/22 10:37:08 - INFO :
         _____________________________________________________
        | preprocess:                                           |
        |   (x - mean) * scale                                  |
        '-------------------------------------------------------'
  config Preprocess args :
        resize_dims           : [640, 640]
        keep_aspect_ratio     : True
        keep_ratio_mode       : letterbox
        pad_value             : 0
        pad_type              : center
        --------------------------
        mean                  : [0.0, 0.0, 0.0]
        scale                 : [1.0, 1.0, 1.0]
        --------------------------
        pixel_format          : rgb
        channel_format        : nhwc

[Running]: tpuc-opt yolo11n.mlir --processor-assign="chip=cv181x mode=INT8 num_device=1 num_core=1 addr_mode=auto" --import-calibration-table="file=yolo11n_calib_table asymmetric=False" --processor-top-optimize --fuse-preprocess="mode=INT8 customization_format=RGB_PACKED align=True" --convert-top-to-tpu=" asymmetric=False doWinograd=False ignore_f16_overflow=False q_group_size=0 matmul_perchannel=False gelu_mode=normal" --canonicalize --weight-fold -o yolo11n_cv181x_int8_sym_tpu.mlir
ConvertUpsampleOp : top.Upsample succeed!
ConvertUpsampleOp : top.Upsample succeed!
Entering FusePreprocessPass.
Inserting PermuteOp.
Inserting LutOp.
PreprocessFuse : tpu.Preprocess succeed!
[Success]: tpuc-opt yolo11n.mlir --processor-assign="chip=cv181x mode=INT8 num_device=1 num_core=1 addr_mode=auto" --import-calibration-table="file=yolo11n_calib_table asymmetric=False" --processor-top-optimize --fuse-preprocess="mode=INT8 customization_format=RGB_PACKED align=True" --convert-top-to-tpu=" asymmetric=False doWinograd=False ignore_f16_overflow=False q_group_size=0 matmul_perchannel=False gelu_mode=normal" --canonicalize --weight-fold -o yolo11n_cv181x_int8_sym_tpu.mlir
[CMD]: model_runner.py --input yolo11n_in_ori.npz --model yolo11n_cv181x_int8_sym_tpu.mlir --output yolo11n_cv181x_int8_sym_tpu_outputs.npz
[                                                  ] 0%root@85cdfc2a6825:/workspace/mywork/workspace/model_yolo11n/Workspace#

**Reply by matsujirushi** [reply]
Please try to remove `--test_input` option. See [here](https://github.com/Seeed-Studio/wiki-documents/discussions/1904#discussioncomment-12671229).

**Comment by Teo-treny** [comment]
Hi,
First, thanks for the guide. I do have a problem tho. I followed the steps (maybe I failed one, I can't really tell) and in NodeRed, when I change the model to my custom one (which I have converted to a cvimodel following the guide), it goes back "YOLO11n Detection" when I reopen the node. Since there are no message error, I can't really tell what is going on. Can someone help me ?

**Reply by litxaohu** [reply]
Hello, I had the same problem some time ago, maybe my experience can help you, you can check the tpu_mlir version in your environment, if it is 1.14 then int8 quantized model will have nothing output. You can download tpu_mlir==1.7 and recompile the project to fix this problem. In version 1.14 F16 precision models are available. You can try them all.

**Comment by manusthep01** [comment]
How to train model for object detection with reCamera?   please advise me.

**Reply by nfs0619** [reply]
Hi! reCamera is for deploying object detection models, not for training them. You can train your model on a PC or in the cloud, then deploy it to reCamera. If you encounter any issues, feel free to contact our tech support at [techsupport@seeed.io](mailto:techsupport@seeed.io).

**Comment by manusthep01** [comment]
Thank for reply to me.
I would like to you advise how to train model step by step for reCamera.

ในวันที่ ศ. 11 ก.ค. 2025 16:20 น. Frank Nie ***@***.***>
เขียนว่า:

**Reply by ReKST-117** [reply]
Hi, we don't have the tutorial directly about model training. The official tutorials from Roboflow and Ultralytics are available, but we cannot guarantee that the model trained by the user can run directly on the ReCamera. Many parameters need to be configured. The link below might help you:

Roboflow tutorial: https://blog.roboflow.com/getting-started-with-roboflow/#quickstart-tutorial-6-minutes
Ultralytics YOLO training tutorial: https://docs.ultralytics.com/modes/train/#introduction

**Comment by manusthep01** [comment]
If I use Ubuntu 20.04 instead of Ubuntu version 22.04, will it still work?"

**Reply by ReKST-117** [reply]
I am not sure what do you mean by still work, but both 20.04 and 22.04 are not latest OS version so should be fine.

**Comment by pooh8rw** [comment]
How do i save the model onto the device so that everytime it reboots it has the custom model?

**Reply by ReKST-117** [reply]
Hi, you might follow this instruction: https://wiki.seeedstudio.com/recamera_develop_with_node-red/#model-selection Train your own model first, converting it to specified format and deploy. But, I don't think you could store your model on recamera and make the custom model a default setup.
